packages/react-i18n/src/index.tsx

```diff
@@ -2,39 +2,40 @@
  * External dependencies
  */
 import * as React from 'react';
-import { __, _n, _nx, _x } from '@wordpress/i18n';
+import memize from 'memize';
+import { I18n, LocaleData } from './i18n-locale';
 import { createHigherOrderComponent } from '@wordpress/compose';
 
-const DEFAULT_LOCALE = 'en';
-
-export interface I18nProps {
-	__: typeof __;
-	_n: typeof _n;
-	_nx: typeof _nx;
-	_x: typeof _x;
-	i18nLocale: string;
+export interface I18nReact {
+	i18nlocale: string;
+	__: I18n[ '__' ];
+	_n: I18n[ '_n' ];
+	_nx: I18n[ '_nx' ];
+	_x: I18n[ '_x' ];
 }
 
-const I18nContext = React.createContext< I18nProps >( makeI18n( DEFAULT_LOCALE ) );
+const DEFAULT_LOCALE = 'en';
+const I18nContext = React.createContext< I18nReact >( makeContextValue( DEFAULT_LOCALE, {} ) );
 
 interface Props {
-	/**
-	 * The current locale
-	 */
-	locale: string;
+	locale?: string;
+	localeData?: LocaleData;
 }
-export const I18nProvider: React.FunctionComponent< Props > = ( { children, locale } ) => {
-	const [ contextValue, setContextValue ] = React.useState< I18nProps >( makeI18n( locale ) );
-	React.useEffect( () => {
-		setContextValue( makeI18n( locale ) );
-	}, [ locale ] );
-	return <I18nContext.Provider value={ contextValue }>{ children }</I18nContext.Provider>;
+export const I18nProvider: React.FunctionComponent< Props > = ( {
+	children,
+	locale = 'en',
+	localeData,
+} ) => {
+	const makeStableContext = React.useRef< typeof makeContextValue >(
+		memize( makeContextValue, { maxSize: 1 } )
+	);
+	return (
+		<I18nContext.Provider value={ makeStableContext.current( locale, localeData ) }>
```

Comment:
Do we need to create a new instance of the `I18n` class every time the locale props change? We can't use the `i18n.setLocaleData()` method to change the locale on the `i18n` instance?

Reply:
Initially I intended to do as you suggest, but two things became apparent.

First, it's very hard to get that right in the react lifecycle.
Second, and related to the first, that doesn't map well to the React paradigm.

`setLocaleData` _merges_ data into the locale data. But with a declarative React API, we'd expect the `localeData` in the provider to map exactly to the locale data. It would be surprising to render `<Provider localeData={…}/>` and not see the localeData reflected exactly in the context!

This seems to align well with the "throw it away and start fresh" paradigm in React. I _did_ try to make progress with a `setLocaleData` approach, but the more I reflect, the more sense a fresh instance every time makes sense to me.